src/libraries/System.Formats.Asn1/src/System/Formats/Asn1/AsnWriter.cs

```diff
@@ -20,6 +20,7 @@ public sealed partial class AsnWriter
         private byte[] _buffer = null!;
         private int _offset;
         private Stack<StackFrame>? _nestingStack;
+        private readonly int? _initialCapacity;
```

Comment:
Instead of adding a new field and doing the alloc lazily, we could just do `_buffer = new byte[initialCapacity]` in the constructor.

I chose not to do it this way so the allocation won't be wasted if the initial capacity is a bad guess.

Reply:
Well, comparing it to `List<T>`:

* initialCapacity:0 is allowed, and means "do whatever you would have done if I called the other overload"
* initialCapacity>0 allocates right away

And the implementation does seem easier if it just allocates right away.

The zero value part I probably care about.  The eager or lazy allocation I don't have a strong opinion on... if you can even think "I might have a strong opinion" then that's good enough for me to justify the lazy approach (the extra if will "hurt" perf, but probably not in any significant sense)

Reply:
Hm... addressed.